src/core/cmp/CoolUtils.py

```diff
@@ -14,29 +17,40 @@ class DeclarationNode(Node):
     pass
 
 class ClassDeclarationNode(DeclarationNode):
-    def __init__(self, idx, features, parent='Object'):
-        self.id = idx
-        self.parent = parent
+    def __init__(self, idx, features, parent=None):
+        self.id = idx.lex
+        self.tid = idx
         self.features = features
+        if not parent:
+            parent = Token("Object", "type")
+            parent.row = idx.row
+            parent.column = idx.column
+        self.parent = parent.lex
+        self.tparent = parent
 
 class AttrDeclarationNode(DeclarationNode):
-    def __init__(self, idx, typex, expr=None):
-        self.id = idx
-        self.type = typex
+    def __init__(self, idx, typex, expr=None, arrow=empty_token):
+        self.id = idx.lex
+        self.tid = idx
+        self.type = typex.lex
+        self.ttype = typex
+        self.arrow = arrow
         self.expr = expr
 
 class FuncDeclarationNode(DeclarationNode):
     def __init__(self, idx, params, return_type, body):
-        self.id = idx
+        self.id = idx.lex
+        self.tid = idx
         self.params = params
-        self.type = return_type
+        self.type = return_type.lex
+        self.ttype = return_type
         self.body = body
 
 class ExpressionNode(Node):
     pass
 
 class IfThenElseNode(ExpressionNode):
-    def __init__(self, condition, if_body, else_body):
+    def __init__(self, condition, if_body, else_body=None):
```

Comment:
```suggestion
    def __init__(self, condition, if_body, else_body):
```
Given that an expression for else's body must exist, I think it would be safer is we don't allow it to be `None`

Reply:
The current grammar allow that behavior, so this change its not necessary right now. It will be solved in the next pull request related to visitors .